lib/web_ui/lib/src/engine/configuration.dart

```diff
@@ -51,14 +52,49 @@ void debugSetConfiguration(FlutterConfiguration configuration) {
 class FlutterConfiguration {
   /// Constructs a configuration from a JavaScript object containing
   /// runtime-supplied properties.
-  FlutterConfiguration(this._js);
+  FlutterConfiguration(JsFlutterConfiguration? config) {
+    addConfigurationOverrides(config);
+  }
 
-  final JsFlutterConfiguration? _js;
+  final List<JsFlutterConfiguration> _configurations = <JsFlutterConfiguration>[];
+
+  /// Adds a configuration [overrides] object that will take precedence over
+  /// any existing configuration.
+  ///
+  /// This is used so we support multiple sources of configuration with a consistent
+  /// precedence. From lowest to highest:
+  ///
+  /// * default static values
+  /// * window.flutterConfiguration object
+  /// * engineInitializer.initEngine(runtimeConfig)
+  ///
+  /// Overrides are always added at position [0] of the array, so the order in
+  /// which this method is called (first with flutterConfiguration), then with
+  /// the runtimeConfig, matters.
+  void addConfigurationOverrides(JsFlutterConfiguration? overrides) {
+    if (overrides != null) {
+      _configurations.insert(0, overrides);
+    }
+  }
+
+  T getConfigValue<T>(T? Function(JsFlutterConfiguration) extractor, T defaultValue) {
```

Comment:
We should assume that some config attributes are accessed frequently, such as `debugShowSemanticsNodes` and `canvasKitMaximuSurfaces`. Therefore, we should aim for this function to be very fast, or avoid calling it repeatedly, e.g. by memoizing the return value at the call side. We should aim to be close to the speed of field access. As it is, it relies on iteration, exceptions, closures, and repeated extraction of values, probably making it 10000x slower than field access.

But also, as I mention in the previous comment, perhaps it's better to not support the merging of multiple config objects, but use either `window.flutterConfiguration` (for legacy reasons) or the shiny new runtime config API.

Reply:
You're right, I wasn't sure if any of these were used in a tight render loop. I've moved all the asserting/warning to the initialization code, so the retrieval is as fast as it can be (I think I actually reverted most of my changes wrt retrieving data, other than renaming the internal field).